File: app/src/main/java/com/e/myfavoritemovies/utils/NetworkUtils.java
package com.e.myfavoritemovies.utils;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import com.e.myfavoritemovies.R;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Scanner;

public class NetworkUtils {

    private static final String TAG = NetworkUtils.class.getSimpleName();

   // private static final String BASE_URL = "http://api.themoviedb.org/3/movie/"; //TODO: move to properties file

    //private static final String API_KEY_VALUE = "2830d7b62b683f28c1c092afbea0196e"; //TODO:move to properties file

    private static final String API_KEY_NAME="api_key";


    /**
     * https://api.themoviedb.org/3/movie/4
     * @param context
     * @param movieType
     * @param page
     * @param favoriteMovies
     * @param id
     * @return
     */
    public static URL buildUrl(Context context, String movieType, int page, boolean favoriteMovies, String id){

        Uri builtUri=null;

         Uri.Builder builder = Uri.parse(context.getString(R.string.base_url)).buildUpon();

             if(!favoriteMovies) {

                 builtUri = builder.appendPath(movieType)
                         .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value)).
                                 appendQueryParameter("page", String.valueOf(page)).build();
             } else {
                 builtUri = builder.appendPath(id)
                         .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value)).build();
             }

        URL url=null;

        try {
            url = new URL(builtUri.toString());
        } catch (MalformedURLException e) {
            e.printStackTrace(); //TODO:remove
        }
        Log.v(TAG, "Built Url "+url);

        return url;
    }

    /**
     * Returns the url of either the popular or top rated movies
     * @param movieType either popular or top rated
     * @return a url for the popular or top rated movies
     */
    /**
     *
     * @param context
     * @param movieId
     * @param page
     * @param favoriteMovies
     * @param id
     * @return
     */
    public static URL buildReviewsUrl(Context context, String movieId, int page, boolean favoriteMovies, String id){

        Uri builtUri=null;

        Uri.Builder builder = Uri.parse(context.getString(R.string.base_url)).buildUpon();


        builtUri = builder.appendPath(movieId).appendPath("reviews")
                    .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value)).
                            appendQueryParameter("page", String.valueOf(page)).build();

        URL url=null;

        try {
            url = new URL(builtUri.toString());
        } catch (MalformedURLException e) {
            e.printStackTrace(); //TODO:remove
        }
        Log.v(TAG, "Built Url "+url);

        return url;
    }

    /**
     * This method returns the entire result from the HTTP response.
     *
     * @param url The URL to fetch the HTTP response from.
     * @return The contents of the HTTP response.
     * @throws IOException Related to network and stream reading
     */
    public static String getResponseFromHttpUrl(URL url) throws IOException {
        HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
        try {
            InputStream in = urlConnection.getInputStream();

            Scanner scanner = new Scanner(in);
            scanner.useDelimiter("\\A");

            boolean hasInput = scanner.hasNext();
            if (hasInput) {
                return scanner.next();
            } else {
                return null;
            }
        } finally {
            urlConnection.disconnect();
        }
    }
}

File: app/src/main/java/com/e/myfavoritemovies/utils/NetworkUtils.kt
package com.e.myfavoritemovies.utils

import android.content.Context
import android.net.Uri
import android.util.Log
import com.e.myfavoritemovies.R
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.Scanner

object NetworkUtils {

    private val TAG: String = NetworkUtils::class.java.simpleName
    private const val API_KEY_NAME = "api_key"

    @JvmStatic
    fun buildUrl(context: Context, movieType: String, page: Int, favoriteMovies: Boolean, id: String?): URL? {
        val builder = Uri.parse(context.getString(R.string.base_url)).buildUpon()

        val builtUri = if (!favoriteMovies) {
            builder.appendPath(movieType)
                .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value))
                .appendQueryParameter("page", page.toString())
                .build()
        } else {
            builder.appendPath(id)
                .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value))
                .build()
        }

        var url: URL? = null
        try {
            url = URL(builtUri.toString())
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        Log.v(TAG, "Built Url $url")
        return url
    }

    @JvmStatic
    fun buildReviewsUrl(context: Context, movieId: String, page: Int, favoriteMovies: Boolean, id: String?): URL? {
        val builder = Uri.parse(context.getString(R.string.base_url)).buildUpon()

        val builtUri = builder.appendPath(movieId).appendPath("reviews")
            .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value))
            .appendQueryParameter("page", page.toString())
            .build()

        var url: URL? = null
        try {
            url = URL(builtUri.toString())
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        Log.v(TAG, "Built Url $url")
        return url
    }

    @JvmStatic
    fun buildTrailersUrl(context: Context, movieId: String, page: Int): URL? {
        val builder = Uri.parse("http://api.themoviedb.org/3/movie/").buildUpon()

        val builtUri = builder.appendPath(movieId).appendPath("videos")
            .appendQueryParameter(API_KEY_NAME, context.getString(R.string.api_key_value))
            .appendQueryParameter("page", page.toString())
            .build()

        var url: URL? = null
        try {
            url = URL(builtUri.toString())
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        Log.v(TAG, "Built Url $url")
        return url
    }

    @JvmStatic
    @Throws(IOException::class)
    fun getResponseFromHttpUrl(url: URL): String? {
        val urlConnection = url.openConnection() as HttpURLConnection
        try {
            val scanner = Scanner(urlConnection.inputStream)
            scanner.useDelimiter("\\A")
            return if (scanner.hasNext()) {
                scanner.next()
            } else {
                null
            }
        } finally {
            urlConnection.disconnect()
        }
    }
}
